javascript/react/common-mistakes-in-react/dependency-array-for-useeffect.md

```diff
@@ -0,0 +1,81 @@
+---
+author: kapnobatai136
+  
+aspects:
+
+  - workout
+
+  - deep
+
+type: normal
+
+category: tip
+
+links:
+  - '[use-deep-compare-effect](https://github.com/kentcdodds/use-deep-compare-effect){website}'
+
+---
+
+# Dependency array for useEffect
+
+---
+## Content
+
+One important feature of the `useEffect()` hook is represented by its dependency array. When included, the dependency array can result in a performance increase due to reducing the number of re-renders. Let's take a look at an example:
+
+```js
+function App() {
+  const [name, setName] = useState("");
+  useEffect(() => {
+    document.title = `Welcome to ${name}!`;
+  }, [name]);
+}
+```
+
+Now, React checks if the current value of name is different from the previous value of name, and only re-renders if that is true.
+
+Let's take a look at another example:
+
+```js
+function App() {
+  const features = ["feature1", "feature2"];
+
+  useEffect(() => {
+    // callback
+  }, [features]);
+}
+```
+
+In this case, React only stores a reference to the `features` array and compares it to the previous reference of the array. We have define the `features` array inside our component, meaning that after every render we recreate the array. When React compares the two references, it will always result on running the callback function. This is because we recreate the `features` array on every render which results in recreating the reference that leads to said array.
```

Comment:
Maybe we can rephrase this one a bit to explain that React uses [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) (and add a footnote for it).

Here's an idea:

```suggestion
It's important to understand that React uses a similar comparison mechanism[1] to `===` when determining if the value of a dependency changed.

In the case of objects and object-like structures such as arrays, `===` will compare the references of two objects, not the data they contain.

From this follows that, even if we pass `[1, 2, 3]` to `useEffect` twice, the actual values within the array will not matter, React will only be interested if the array itself (i.e. the reference to it) is the same, not its content.

In the example above we are creating the `features` array within the component `App`. This means that anytime `App` renders, `features` will contain a new array instance of with the same contents. When React checks if `features` array is the same as before, it never will be because we're recreating it on each render, therefore the `useEffect` hook will always run.
```